fedimintd/src/bin/main.rs

```diff
@@ -72,7 +75,43 @@ async fn main() -> anyhow::Result<()> {
         registry.init();
     }
 
-    let mut task_group = TaskGroup::new();
+    let mut root_task_group = TaskGroup::new();
+    root_task_group.install_kill_handler();
+
+    let task_group = root_task_group.clone();
+    let (main_shutdown_tx, main_shutdown_rx) = tokio::sync::oneshot::channel::<()>();
+
+    root_task_group
+        .spawn_local("main", move |_task_handle| async move {
+            match run(opts, task_group).await {
+                Ok(()) => {
+                    info!("Shutting down");
+                }
+                Err(e) => {
+                    error!(?e, "Main task returned error, shutting down");
+                }
+            }
+
+            main_shutdown_tx
```

Comment:
Seems like we should `task_group.shutdown();`here. (If we are out of run, we want everything else to finish).

Reply:
So we wouldn't need  `main_shutdown_rx.await` (or that `oneshot` channel at all)` and we could just do `root_task_group.join_all().await`

Reply:
Oh... This is bad.

Possibly we should have a `while` loop there and lock, pop one, wait one it, until we don't have any more?

Reply:
I wouldn't be surprised if this wasn't one of the root issues. Good catch.

Who wrote it? What was he thinking. :D

Reply:
I'll take a look tomorrow with a fresh mind. The while loop seems ok, we'd want a `VecDequeue` though to pop from the opposite side than where we push. Since we don't seem to do random access we might as well use an unbounded channel.

Reply:
I ended up making it a `VecDequeue` since the channel would have meant two member vars and a mutex around the receiver still. We should come back and refactor the `TaskGroup` such that handles can spawn, `TaskGroup` is not clone anymore and is the only thing that can `join_all`. That way consuming `self` would actually make sense (when it's just an `Arc` that can be cloned consuming self means ~nothing).